core/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -94,6 +94,22 @@
     private final SetOnce<Set<String>> firstLevelNames = new SetOnce<>();
 
     /**
+     * The set of deprecated settings tracked by this settings object.
+     */
+    private final Set<String> deprecatedSettings = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+    /**
+     * Add the key as a tracked deprecated setting.
+     *
+     * @param key the key of the deprecated setting to track
+     * @return true if the setting was not already tracked as a deprecated setting, otherwise false
+     */
+    boolean addDeprecatedSetting(final String key) {
+        assert settings.containsKey(key);
```

Comment:
Can you use `exists` for this assert? Otherwise it won't work for deprecated secure settings (which don't exist in the `settings` map.